WIP Lexer.c

TheGreatSoider opened the pull request:
Adjusting it for norminette creates problems

TheGreatSoider committed:
WIP Lexer.c

TheGreatSoider committed:
got the quotes to work.

TheGreatSoider committed:
WIP Lexer.

Changed approach from linked list idea to pointers to array

TheGreatSoider committed:
can handle quotes now.

This "works as" intended now.
This
"works
as"
intended
now.
->
This
"works as"
intended
now.

zklim commented on lexer/lexer.c:
```diff
@@ -22,23 +22,28 @@
 
-
-typedef enum s_token
+typedef struct s_quotes
```
Why are all header definitions here?

zklim commented on lexer/lexer.c:
```diff
@@ -56,87 +61,223 @@ bool	streq(char *str1, char *str2)
+	return (ptr);
+}
+
+// char	**adjust_tokens(char **lexer_tokens)
```
What are these commented codes for?

zklim commented on lexer/lexer.c:
```diff
@@ -56,87 +61,223 @@ bool	streq(char *str1, char *str2)
+int main() {
+	char input[256];
+	printf("Enter a command: ");
+	fgets(input, 256, stdin);
```
We should be using readline instead of fgets

TheGreatSoider replied:
this is just for testing. not for the final product.

zklim commented on lexer/lexer.c:
```diff
@@ -56,87 +61,223 @@ bool	streq(char *str1, char *str2)
+	char input[256];
+	printf("Enter a command: ");
+	fgets(input, 256, stdin);
+	input[strcspn(input, "\n")] = 0;  // Remove trailing newline
```
I don't think we can do this kind of declaration either

TheGreatSoider replied:
just for testing purposes

zklim commented on lexer/lexer.c:
```diff
@@ -56,87 +61,223 @@ bool	streq(char *str1, char *str2)
 
-// 	return 0;
-// }
+	for (int i = 0; lexer_tokens[i] != NULL; i++) {
```
for loops are forbidden

TheGreatSoider replied:
I know. ease for testing purposes

zklim commented on lexer/lexer.c:
```diff
@@ -22,23 +22,28 @@
 
 char	*ft_strtrim(char const *s1, char const *set);
 char	*ft_substr(char const *s, unsigned int start, size_t len);
+t_parser_token	*parser(char **lexer_tokens);
```
Unknown type
```shell
lexer.c:25:1: error: unknown type name 't_parser_token'
t_parser_token  *parser(char **lexer_tokens);
^
1 error generated.
```

zklim reviewed (changes requested):
Why are there 3 different lexer files?
I encountered a lot of errors during compiling too.

TheGreatSoider committed:
parser can identify redirects and commands.

can identify, now just need to go through functions to execute.

TheGreatSoider committed:
Delete lexer/lexer2.c

TheGreatSoider committed:
Delete lexer/a.out

TheGreatSoider committed:
Delete lexer/lexer_WIP.c

TheGreatSoider committed:
Delete lexer.c

TheGreatSoider committed:
moved them to src directory

TheGreatSoider committed:
Parser and lexer

Lexer fixed a bug.

Parser together with executor. Will separate later. Just for ease of testing atm.

Could get pwd to work. Everything else not yet.

The idea is it reads the first exe_tok-> cmd  which would be a one of the commands (echo, pwd, etc). It will then increment and start reading the rest as the args as it increment.

TheGreatSoider committed:
Tidied up and added include files

Tidied up and added includes file for relevant files.

Now just need to adjust built-ins accordingly.

TheGreatSoider committed:
Saving progress. 

Doing the cmdpath now. Made some changes to inbuilts to work with the execute. 

Just need to tidy the work up and finish cmdpath. 

Lastly, get the free in check.

zklim commented on src/executor/executor.c:
```diff
@@ -12,103 +12,287 @@
+	{
+		path_len = path_end - path_start;
+		copy_path(abs_cmd_path, path_start, path_len, cmd);
+		if (access(abs_cmd_path, X_OK) == 0)
```
What access() does? it's not allowed external function

zklim replied:
Ok it's allowed

zklim commented on src/executor/executor.c:
```diff
@@ -12,103 +12,287 @@
+	int		path_len;
+
+	path_start = getenv("PATH");
+	abs_cmd_path = malloc(PATH_MAX);
```
PATH_MAX not defined and this is not the right way to do malloc

zklim reviewed (changes requested):
Can't build and test this changes

TheGreatSoider committed:
Added cmd_path and changes to executor.

Added cmd_path.c to test if it works. 

Changed executor because previously it only accepts the first argument. I.e. echo 123 213. only returns 123. 

The updated should return 123 213 but now has compiling error.

Ignore the init function as I will remove it later and malloc accordingly. Right now I leave it for convenience.

May need help with pipes cause I am struggling to understand how to handle more than 1 pipe.

zklim reviewed (commented):
How to test these new codes? I need to understand how they work before proceeding to next steps

TheGreatSoider committed:
Correct executor file

Should be able to compile. Segfaults after executing. Figuring out why.

To compile cmd_path use -> gcc cmd_path.c ../../libft/ft_strlen.c ../../libft/ft_strjoin.c ../../libft/ft_split.c ../../libft/ft_strdup.c

TheGreatSoider committed:
added latest cmd_path.c

forgot to add cmd_path.c to test.

TheGreatSoider committed:
Moved functions and add functions

Moved some functions to expander file. Made an executor copy to test new solution. 

Can execute echo, pwd, env, unset, and export. But always will segfault or have a double free warning. Looking into it now.

Once that is fixed, all that is left is pipe.

TheGreatSoider committed:
Update expander.c

TheGreatSoider committed:
Added envp

added envp to store and handle envp variable

TheGreatSoider committed:
Changed and added lexer

change lexer to fix vulnerabilities and segfault issues.

moved functions to files to satisfy norm

TheGreatSoider committed:
Added files to satisfy norm.

Changed parser slightly to fix vulnerabilities and segfault.

Added files to satisfy norm.

TheGreatSoider committed:
Updated and added function for expander

To fix segfault, later will fix for norm

TheGreatSoider committed:
Finished piping operations

Added and organized functions into files

TheGreatSoider committed:
up to date includes

TheGreatSoider committed:
Delete src/executor/executor_copy.c

Not needed anymore
